File: src/asr/utils/ogg-muxer.ts
/**
 * 流式 OGG Muxer
 *
 * 将 Opus 数据包流式封装为 OGG 流
 * 纯 JavaScript 实现，无需额外依赖
 *
 * @module ogg-muxer
 */

import { Buffer } from 'node:buffer';

/**
 * CRC32 查找表（OGG 使用的多项式: 0x04c11db7）
 */
const CRC32_TABLE: Uint32Array = (() => {
  const table = new Uint32Array(256);
  for (let i = 0; i < 256; i++) {
    let r = i << 24;
    for (let j = 0; j < 8; j++) {
      r = (r & 0x80000000) !== 0 ? (r << 1) ^ 0x04c11db7 : r << 1;
    }
    table[i] = r >>> 0;
  }
  return table;
})();

/**
 * 计算 OGG CRC32 校验和
 * @param data 数据缓冲区
 * @returns CRC32 值
 */
function crc32(data: Uint8Array): number {
  let crc = 0;
  for (let i = 0; i < data.length; i++) {
    crc = ((crc << 8) ^ CRC32_TABLE[((crc >>> 24) ^ data[i]) & 0xff]) >>> 0;
  }
  return crc;
}

/**
 * 创建 OpusHead 包
 * @param sampleRate 采样率
 * @param channels 声道数
 * @param preSkip 预跳过采样数
 * @returns OpusHead Buffer
 */
function createOpusHead(sampleRate: number, channels: number, preSkip: number): Buffer {
  const buffer = Buffer.alloc(19);

  // "OpusHead" 魔数 (8 bytes)
  buffer.write('OpusHead', 0, 'ascii');

  // 版本号 (1 byte) - 总是 1
  buffer.writeUInt8(1, 8);

  // 声道数 (1 byte)
  buffer.writeUInt8(channels, 9);

  // Pre-skip (2 bytes, little-endian)
  buffer.writeUInt16LE(preSkip, 10);

  // 采样率 (4 bytes, little-endian)
  buffer.writeUInt32LE(sampleRate, 12);

  // Output gain (2 bytes, little-endian) - 总是 0
  buffer.writeUInt16LE(0, 16);

  // Channel mapping family (1 byte) - 0 = 单声道或立体声
  buffer.writeUInt8(0, 18);

  return buffer;
}

/**
 * 创建 OpusTags 包
 * @param encoder 编码器名称
 * @returns OpusTags Buffer
 */
function createOpusTags(encoder: string): Buffer {
  const encoderTag = `encoder=${encoder}`;
  const encoderTagBuffer = Buffer.from(encoderTag, 'utf-8');

  // OpusTags 结构:
  // "OpusTags" (8) + Vendor Length (4) + Vendor String + Tag Count (4) + Tags...
  const vendorBuffer = Buffer.from('univoice', 'utf-8');

  // 总大小: 8 + 4 + vendor.length + 4 + 4 + tag.length
  const totalSize = 8 + 4 + vendorBuffer.length + 4 + 4 + encoderTagBuffer.length;
  const buffer = Buffer.alloc(totalSize);

  let offset = 0;

  // "OpusTags" 魔数 (8 bytes)
  buffer.write('OpusTags', offset, 'ascii');
  offset += 8;

  // Vendor string length (4 bytes, little-endian)
  buffer.writeUInt32LE(vendorBuffer.length, offset);
  offset += 4;

  // Vendor string
  vendorBuffer.copy(buffer, offset);
  offset += vendorBuffer.length;

  // Tag count (4 bytes, little-endian)
  buffer.writeUInt32LE(1, offset);
  offset += 4;

  // Tag length (4 bytes, little-endian)
  buffer.writeUInt32LE(encoderTagBuffer.length, offset);
  offset += 4;

  // Tag string
  encoderTagBuffer.copy(buffer, offset);

  return buffer;
}

/**
 * 创建 OGG 页面
 * @param packets 数据包数组（一个页面可包含多个包）
 * @param pageSequence 页面序列号
 * @param granulePosition 颗粒位置（采样数）
 * @param serialNumber 比特流序列号
 * @param flags 页面标志
 * @returns OGG 页面 Buffer
 */
function createOggPage(
  packets: Buffer[],
  pageSequence: number,
  granulePosition: bigint,
  serialNumber: number,
  flags: { bos: boolean; eos: boolean }
): Buffer {
  // 计算总数据大小
  const totalDataSize = packets.reduce((sum, p) => sum + p.length, 0);

  // 计算段表
  const segments: number[] = [];
  for (const packet of packets) {
    let remaining = packet.length;
    while (remaining >= 255) {
      segments.push(255);
      remaining -= 255;
    }
    segments.push(remaining);
  }

  // OGG 页面头部大小: 27 + segments.length
  const headerSize = 27 + segments.length;
  const pageSize = headerSize + totalDataSize;

  // 创建页面缓冲区（先不包含 CRC）
  const page = Buffer.alloc(pageSize);
  let offset = 0;

  // OggS 捕获模式 (4 bytes)
  page.write('OggS', offset, 'ascii');
  offset += 4;

  // 版本 (1 byte) - 总是 0
  page.writeUInt8(0, offset);
  offset += 1;

  // 头类型标志 (1 byte)
  // bit 0: continued packet
  // bit 1: BOS (beginning of stream)
  // bit 2: EOS (end of stream)
  let headerType = 0;
  if (flags.bos) headerType |= 0x02;
  if (flags.eos) headerType |= 0x04;
  page.writeUInt8(headerType, offset);
  offset += 1;

  // 颗粒位置 (8 bytes, little-endian)
  page.writeBigUInt64LE(granulePosition, offset);
  offset += 8;

  // 比特流序列号 (4 bytes, little-endian)
  page.writeUInt32LE(serialNumber, offset);
  offset += 4;

  // 页面序列号 (4 bytes, little-endian)
  page.writeUInt32LE(pageSequence, offset);
  offset += 4;

  // CRC32 (4 bytes) - 先填 0，后面计算
  const crcOffset = offset;
  page.writeUInt32LE(0, offset);
  offset += 4;

  // 段数 (1 byte)
  page.writeUInt8(segments.length, offset);
  offset += 1;

  // 段长度表
  for (const seg of segments) {
    page.writeUInt8(seg, offset);
    offset += 1;
  }

  // 数据段
  for (const packet of packets) {
    packet.copy(page, offset);
    offset += packet.length;
  }

  // 计算 CRC32 并填入
  const crcValue = crc32(page);
  page.writeUInt32LE(crcValue, crcOffset);

  return page;
}

/**
 * OGG Muxer 选项
 */
export interface OggMuxerOptions {
  /** 采样率，默认 16000 */
  sampleRate?: number;
  /** 声道数，默认 1 */
  channels?: number;
  /** Opus 帧大小（毫秒），默认 60 */
  frameSizeMs?: number;
  /** 编码器名称，默认 'univoice' */
  encoder?: string;
}

/**
 * 创建流式 OGG Muxer
 *
 * 将 Opus 数据包流式封装为 OGG 流。每收到一个 Opus packet，立即 yield 一个 OGG page。
 *
 * @param opusPackets Opus 数据包流
 * @param options 选项配置
 * @returns OGG 页面流
 *
 * @example
 * ```typescript
 * const oggStream = createOggMuxer(hardwareStream, { sampleRate: 16000 });
 * for await (const oggPage of oggStream) {
 *   // 处理每个 OGG 页面
 * }
 * ```
 */
export async function* createOggMuxer(
  opusPackets: AsyncIterable<Buffer>,
  options?: OggMuxerOptions
): AsyncIterable<Buffer> {
  const {
    sampleRate = 16000,
    channels = 1,
    frameSizeMs = 60,
    encoder = 'univoice',
  } = options || {};

  // 生成随机序列号
  const serialNumber = Math.floor(Math.random() * 0xffffffff);

  // Opus 内部采样率总是 48000Hz
  // pre-skip 推荐值为 312 (for 48kHz)
  // 参考: https://tools.ietf.org/html/rfc7845#section-5.1
  const preSkip = 312;

  // Opus 内部采样率总是 48000Hz，颗粒位置使用 48000Hz 计算
  // 参考: https://tools.ietf.org/html/rfc7845#section-3
  const opusInternalSampleRate = 48000;
  const samplesPerFrame = (opusInternalSampleRate / 1000) * frameSizeMs;

  // 页面序列号
  let pageSequence = 0;

  // 1. yield OpusHead 页面 (BOS - Beginning Of Stream)
  const opusHead = createOpusHead(sampleRate, channels, preSkip);
  yield createOggPage([opusHead], pageSequence++, 0n, serialNumber, { bos: true, eos: false });

  // 2. yield OpusTags 页面
  const opusTags = createOpusTags(encoder);
  yield createOggPage([opusTags], pageSequence++, 0n, serialNumber, { bos: false, eos: false });

  // 3. 流式处理 Opus 数据包
  // 颗粒位置从 pre-skip 开始
  let granulePosition = BigInt(preSkip);

  for await (const packet of opusPackets) {
    const data = Buffer.isBuffer(packet) ? packet : Buffer.from(packet);

    // 更新颗粒位置
    granulePosition += BigInt(samplesPerFrame);

    // yield OGG 页面（中间页面不设置 EOS 标志）
    yield createOggPage([data], pageSequence++, granulePosition, serialNumber, {
      bos: false,
      eos: false,
    });
  }

  // 注意：流式场景下，最后一个页面的 EOS 标志需要单独处理
  // 调用方可以在结束时调用 createEosPage 或让 ASR 服务通过结束信号处理
}

/**
 * 创建 EOS (End Of Stream) 页面
 *
 * 用于标记 OGG 流的结束
 *
 * @param serialNumber 比特流序列号
 * @param pageSequence 下一个页面序列号
 * @param granulePosition 最后的颗粒位置
 * @returns 带有 EOS 标志的 OGG 页面
 */
export function createEosPage(
  serialNumber: number,
  pageSequence: number,
  granulePosition: bigint
): Buffer {
  // 创建一个空的 EOS 页面
  return createOggPage([], pageSequence, granulePosition, serialNumber, {
    bos: false,
    eos: true,
  });
}

/**
 * 创建完整的 OGG Muxer 实例
 *
 * 提供更多控制能力，包括：
 * - 获取序列号
 * - 手动发送 EOS 页面
 */
export interface OggMuxer {
  /** OGG 流 */
  stream: AsyncIterable<Buffer>;
  /** 比特流序列号 */
  serialNumber: number;
  /** 发送 EOS 标记并结束流 */
  finish: () => void;
}

/**
 * 创建可控的 OGG Muxer 实例
 *
 * 提供更精细的控制，可以在流结束时发送 EOS 标记
 *
 * @param opusPackets Opus 数据包流
 * @param options 选项配置
 * @returns OGG Muxer 实例
 */
export function createOggMuxerWithEos(
  opusPackets: AsyncIterable<Buffer>,
  options?: OggMuxerOptions
): OggMuxer {
  const {
    sampleRate = 16000,
    channels = 1,
    frameSizeMs = 60,
    encoder = 'univoice',
  } = options || {};

  // 生成随机序列号
  const serialNumber = Math.floor(Math.random() * 0xffffffff);

  // Opus 参数
  const preSkip = 312;
  const opusInternalSampleRate = 48000;
  const samplesPerFrame = (opusInternalSampleRate / 1000) * frameSizeMs;

  // 状态
  let pageSequence = 0;
  let granulePosition = BigInt(preSkip);
  let finished = false;

  // 创建流
  async function* stream(): AsyncIterable<Buffer> {
    // OpusHead
    const opusHead = createOpusHead(sampleRate, channels, preSkip);
    yield createOggPage([opusHead], pageSequence++, 0n, serialNumber, { bos: true, eos: false });

    // OpusTags
    const opusTags = createOpusTags(encoder);
    yield createOggPage([opusTags], pageSequence++, 0n, serialNumber, { bos: false, eos: false });

    // 音频数据包
    for await (const packet of opusPackets) {
      if (finished) break;

      const data = Buffer.isBuffer(packet) ? packet : Buffer.from(packet);
      granulePosition += BigInt(samplesPerFrame);

      yield createOggPage([data], pageSequence++, granulePosition, serialNumber, {
        bos: false,
        eos: false,
      });
    }

    // 发送 EOS 页面（如果已标记结束）
    if (finished) {
      yield createEosPage(serialNumber, pageSequence, granulePosition);
    }
  }

  return {
    stream: stream(),
    serialNumber,
    finish: () => {
      finished = true;
    },
  };
}

File: src/asr/utils/ogg_muxer.rs
/// OGG 封装器 - 将裸 Opus 数据包封装为 OGG 流
///
/// OGG 页面格式:
/// - 页眉 (27 bytes): OggS + 版本 + 标志 + granule_position + serial + sequence + CRC + segments
/// - 段表: 每个数据包的长度
/// - 数据包内容
///
/// Opus 在 OGG 中的结构:
/// - 第1页: OpusHead (BOS)
/// - 第2页: OpusTags
/// - 后续: 音频数据包
use std::time::{SystemTime, UNIX_EPOCH};

/// OGG CRC32 查找表（前向 CRC32，多项式 0x04C11DB7）
///
/// OGG 规范要求使用前向（MSB-first）CRC32，而非标准的反射 CRC32。
/// 参考: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
const CRC32_TABLE: [u32; 256] = {
    let mut table = [0u32; 256];
    let mut i = 0;
    while i < 256 {
        let mut r = i << 24;
        let mut j = 0;
        while j < 8 {
            if r & 0x8000_0000 != 0 {
                r = (r << 1) ^ 0x04c1_1db7;
            } else {
                r <<= 1;
            }
            j += 1;
        }
        table[i as usize] = r;
        i += 1;
    }
    table
};

/// 计算 OGG 前向 CRC32
fn ogg_crc32(data: &[u8]) -> u32 {
    let mut crc: u32 = 0;
    for &byte in data {
        let idx = (((crc >> 24) ^ (byte as u32)) & 0xFF) as usize;
        crc = (crc << 8) ^ CRC32_TABLE[idx];
    }
    crc
}

/// OGG Muxer 配置选项
#[derive(Debug, Clone)]
pub struct OggMuxerOptions {
    /// 采样率（用于 granule position 计算，默认 16000）
    pub sample_rate: u32,
    /// 声道数（默认 1）
    pub channels: u8,
    /// 每帧时长毫秒（默认 60）
    pub frame_size_ms: u32,
}

impl Default for OggMuxerOptions {
    fn default() -> Self {
        Self {
            sample_rate: 16000,
            channels: 1,
            frame_size_ms: 60,
        }
    }
}

/// OGG Muxer - 将裸 Opus 数据包流式封装为 OGG 页面
pub struct OggMuxer {
    serial_number: u32,
    page_sequence: u32,
    granule_position: u64,
    channels: u8,
    frame_size_ms: u32,
    started: bool,
    finished: bool,
}

impl OggMuxer {
    /// 当前页面序号
    pub fn page_sequence(&self) -> u32 {
        self.page_sequence
    }
}

impl OggMuxer {
    /// 创建新的 OGG Muxer
    pub fn new(options: OggMuxerOptions) -> Self {
        // 生成伪随机 serial number（基于时间戳）
        let serial_number = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_micros() as u32)
            .unwrap_or(42);

        Self {
            serial_number,
            page_sequence: 0,
            // Opus pre-skip: 对于 48kHz Opus 内部采样率，推荐值为 312
            // 参考: RFC 7845 Section 5.1
            granule_position: 312,
            channels: options.channels,
            frame_size_ms: options.frame_size_ms,
            started: false,
            finished: false,
        }
    }

    /// 写入一个 Opus 数据包，返回产生的 OGG 页面列表
    ///
    /// 第一次调用时返回 [OpusHead, OpusTags, 第一帧数据]
    /// 后续调用返回 [数据页面]
    pub fn push_packet(&mut self, opus_packet: &[u8]) -> Vec<Vec<u8>> {
        let mut pages = Vec::new();

        if !self.started {
            self.started = true;
            // 第一个页面：OpusHead（BOS）
            pages.push(self.create_opus_head_page());
            // 第二个页面：OpusTags
            pages.push(self.create_opus_tags_page());
        }

        // 更新 granule position
        // Opus 内部采样率总是 48000Hz，颗粒位置使用 48kHz 计算
        // 参考: RFC 7845 Section 3
        let samples_per_frame = (48000u64 * self.frame_size_ms as u64) / 1000;
        self.granule_position += samples_per_frame;

        // 创建数据页面
        pages.push(self.create_data_page(opus_packet, false));

        pages
    }

    /// 结束流，返回 EOS 页面（如果有数据尚未发送）
    pub fn finish(&mut self) -> Option<Vec<u8>> {
        if self.finished {
            return None;
        }
        self.finished = true;
        // 在 Opus 流结束时，发送一个空页面标记 EOS
        // 但一般 ASR 服务不需要 EOS 页面，流式发送完数据后关闭连接即可
        None
    }

    /// 创建 OpusHead 页面（BOS）
    fn create_opus_head_page(&mut self) -> Vec<u8> {
        let mut packet = Vec::with_capacity(19);

        // OpusHead magic
        packet.extend_from_slice(b"OpusHead");
        // Version
        packet.push(1);
        // Channel count
        packet.push(self.channels);
        // Pre-skip (312 for 48kHz, RFC 7845 Section 5.1)
        packet.extend_from_slice(&312u16.to_le_bytes());
        // Input sample rate (always 48000 for Opus)
        packet.extend_from_slice(&48000u32.to_le_bytes());
        // Output gain (0)
        packet.extend_from_slice(&0u16.to_le_bytes());
        // Mapping family (0 = mono/stereo)
        packet.push(0);

        // 创建页面
        self.create_page(&[&packet], 2) // BOS flag
    }

    /// 创建 OpusTags 页面
    fn create_opus_tags_page(&mut self) -> Vec<u8> {
        let vendor_string = b"univoice-rs";
        let mut packet = Vec::new();

        // OpusTags magic
        packet.extend_from_slice(b"OpusTags");
        // Vendor string length
        packet.extend_from_slice(&(vendor_string.len() as u32).to_le_bytes());
        // Vendor string
        packet.extend_from_slice(vendor_string);
        // User comment list length (0)
        packet.extend_from_slice(&0u32.to_le_bytes());

        self.create_page(&[&packet], 0) // normal flag
    }

    /// 创建数据页面
    fn create_data_page(&mut self, opus_data: &[u8], eos: bool) -> Vec<u8> {
        let flags = if eos { 4 } else { 0 };
        self.create_page(&[opus_data], flags)
    }

    /// 构建 OGG 页面并自动递增页序号
    ///
    /// `segments`: 数据包列表（每个可能跨多个段）
    /// `header_type_flag`: 2=BOS, 4=EOS, 0=normal
    fn create_page(&mut self, packets: &[&[u8]], header_type_flag: u8) -> Vec<u8> {
        // 构建段表（segment table）
        let mut segment_table = Vec::new();
        for &packet in packets {
            // OGG 每段最大 255 字节，超过则分多段
            let mut remaining = packet.len();
            while remaining > 255 {
                segment_table.push(255u8);
                remaining -= 255;
            }
            segment_table.push(remaining as u8);
        }

        let num_segments = segment_table.len() as u8;

        // 预计算页面大小（不含 CRC）
        let page_size = 27 + num_segments as usize + packets.iter().map(|p| p.len()).sum::<usize>();
        let mut page = Vec::with_capacity(page_size + 4);

        // 1. 页眉 (27 bytes)
        page.extend_from_slice(b"OggS"); // capture pattern
        page.push(0); // version
        page.push(header_type_flag); // header_type
        page.extend_from_slice(&self.granule_position.to_le_bytes()); // granule_position
        page.extend_from_slice(&self.serial_number.to_le_bytes()); // serial_number
        page.extend_from_slice(&self.page_sequence.to_le_bytes()); // page_sequence
        // CRC placeholder (4 bytes, will be filled later)
        let crc_offset = page.len();
        page.extend_from_slice(&0u32.to_le_bytes());
        page.push(num_segments); // num_segments

        // 2. 段表
        page.extend_from_slice(&segment_table);

        // 3. 数据包
        for &packet in packets {
            page.extend_from_slice(packet);
        }

        // 4. 计算 CRC (将 CRC 字段设为 0 后计算)
        let crc = ogg_crc32(&page);
        let crc_bytes = crc.to_le_bytes();
        page[crc_offset..crc_offset + 4].copy_from_slice(&crc_bytes);

        // 递增页序号
        self.page_sequence += 1;

        page
    }
}

/// 创建一个从 Opus 数据包流到 OGG 页面流适配器
///
/// 将输入的 Opus 数据包流（每个元素是一个裸 Opus 帧）
/// 转换为 OGG 页面流（每个元素是一个完整的 OGG 页面）
pub fn create_ogg_stream(
    opus_packets: impl futures_util::Stream<Item = Vec<u8>> + Send + 'static,
    options: OggMuxerOptions,
) -> impl futures_util::Stream<Item = Vec<u8>> {
    use futures_util::StreamExt;

    let mut muxer = OggMuxer::new(options);

    // 使用 async_stream 将 Opus 包流映射为 OGG 页面流
    async_stream::stream! {
        tokio::pin!(opus_packets);
        while let Some(packet) = opus_packets.next().await {
            let pages = muxer.push_packet(&packet);
            for page in pages {
                yield page;
            }
        }
        // 流结束时发送 EOS
        if let Some(eos_page) = muxer.finish() {
            yield eos_page;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ogg_muxer_basic() {
        let mut muxer = OggMuxer::new(OggMuxerOptions::default());

        // 第一个 Opus 数据包（模拟一个短包）
        let first_packet = vec![0u8; 64]; // 模拟 Opus 数据
        let pages = muxer.push_packet(&first_packet);

        // 第一次 push 应产生 3 个页面: OpusHead + OpusTags + 数据
        assert_eq!(pages.len(), 3, "第一次 push 应产生 3 个页面");

        // 验证 OpusHead 页面
        let head_page = &pages[0];
        assert!(head_page.len() > 27, "页面应大于页眉大小");
        assert_eq!(&head_page[0..4], b"OggS", "OGG 捕获模式");
        assert_eq!(head_page[5], 2, "BOS 标志");

        // 验证 OpusTags 页面
        let tags_page = &pages[1];
        assert_eq!(&tags_page[0..4], b"OggS", "OGG 捕获模式");
        assert_eq!(tags_page[5], 0, "普通页面");

        // 验证数据页面
        let data_page = &pages[2];
        assert_eq!(&data_page[0..4], b"OggS", "OGG 捕获模式");

        // 第二个数据包
        let second_packet = vec![0u8; 128];
        let pages2 = muxer.push_packet(&second_packet);
        assert_eq!(pages2.len(), 1, "后续 push 应产生 1 个页面");
    }

    #[test]
    fn test_ogg_muxer_crc() {
        let mut muxer = OggMuxer::new(OggMuxerOptions::default());
        let pages = muxer.push_packet(&[0u8; 64]);

        // 验证所有页面的 CRC 不为 0（说明被正确计算了）
        for page in &pages {
            let crc_bytes = &page[22..26];
            let crc = u32::from_le_bytes([crc_bytes[0], crc_bytes[1], crc_bytes[2], crc_bytes[3]]);
            // CRC 极大概率不为 0（即使数据全 0）
            assert_ne!(crc, 0, "CRC 不应为 0");
        }
    }

    #[test]
    fn test_ogg_muxer_sequence() {
        let mut muxer = OggMuxer::new(OggMuxerOptions::default());

        // 第1次: 3 页（BOS + Tags + data0）
        let p0 = muxer.push_packet(&[0u8; 32]);
        assert_eq!(p0.len(), 3);

        // 验证页面序列号
        assert_eq!(u32_from_le_slice(&p0[0][18..22]), 0, "OpusHead 页序号 0");
        assert_eq!(u32_from_le_slice(&p0[1][18..22]), 1, "OpusTags 页序号 1");
        assert_eq!(u32_from_le_slice(&p0[2][18..22]), 2, "第一帧数据页序号 2");

        // 第2次: 1 页（data1）
        let p1 = muxer.push_packet(&[0u8; 48]);
        assert_eq!(p1.len(), 1);
        assert_eq!(u32_from_le_slice(&p1[0][18..22]), 3, "第二帧数据页序号 3");

        // 第3次: 1 页（data2）
        let p2 = muxer.push_packet(&[0u8; 64]);
        assert_eq!(p2.len(), 1);
        assert_eq!(u32_from_le_slice(&p2[0][18..22]), 4, "第三帧数据页序号 4");
    }

    fn u32_from_le_slice(bytes: &[u8]) -> u32 {
        u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]])
    }
}
